### PR TITLE
SM-79: create roles, permissions and synchronisation service

### LINKED ISSUE #79: Define available user roles
## Description

We need to analyse which roles we need within the application at this moment. We can then create these roles automatically in each environment.

### What types of users will live within the current system?

* **Member (default)** lowest role within the application that all standard users have. 
* **Handyman or concierge:** user that is part of the "Klusgroep" and has extra capabilities to do his job
* **Admin:** role that has all permissions

### Technical requirements

These roles and permissions should be synchronised or maintained in each environment. Both `web` and `rest_api` guards needs to have the same setup.

## Advice

Since we don't have a admin dashboard for now where we can create a functionality to manage the roles and permissions within the system, we will be maintaining these mappings in-code for now. This will spare us some time for now and we can make sure the roles and permissions are on each environment equal.

## Design

Create a new directory in `App` called `ACL` (meaning Access Control List) containing interface `PermissionsInterface`. This interface will contain a list of public constants defining all the permissions within the system. 

We will create the roles-permission mapping inside specific role classes in `App/ACL/Roles`. These classes will provide a method called `getPermissions(): array` providing all their permissions.

A service should synchronise these mappings with the database. We can execute this action automatically after each deployment, so new changes to these mappings are applied.

The service can then be used in a console command. This command can be fired after each deployment

#### Feature

* [x] `ACLService@synchroniseRolesAndPermissions` - Make sure the expected state is there after running the logic. Since the logic uses functionality from the `laravel-permission` package, testing this logic is unnecessary and not our responsibility.
* [x] `GuardService@getByName` - Make sure it can find one.
* [x] `GuardService@getByName` - Make sure it throws an exception if nothing is found.
* [x] `GuardService@ existsByName ` - Make sure it can find one.
*  [x] `GuardService@ existsByName ` - Make sure it can not find one. 
* [x] `SynchroniseRoleAndPermissions@handle` - Follow the happy flow.
* [x] `SynchroniseRoleAndPermissions@handle` - Stop executing if guard name is not valid.
* [x] Test the gate fore hook for role admin. It should always have the rights.

### PR DESCRIPTION
**Pull request by cyrildewit**
Closes #79

### CODE REVIEW

File: app/ACL/ACLService.php [hunk 1]
```diff
@@ -0,0 +1,51 @@
...
+    private RolesProvider $rolesProvider;
+
+    public function __construct(RolesProvider $rolesProvider)
+    {
+        $this->rolesProvider = $rolesProvider;
+    }
+
+    public function synchroniseRolesAndPermissions(): void
+    {
+        $guards = new Collection([
```
Comment (cyrildewit) [inline app/ACL/ACLService.php:29]: This list is part of the logic of this class.

File: app/ACL/Roles/Providers/DefaultRolesProvider.php [hunk 2]
```diff
@@ -0,0 +1,24 @@
...
+
+use App\ACL\Contracts\RolesProvider;
+use App\ACL\Roles\ConciergeRole;
+use App\ACL\Roles\MemberRole;
+use App\ACL\Roles\RoleCollection;
+
+/**
+ * @codeCoverageIgnore
+ */
+final class DefaultRolesProvider implements RolesProvider
```
Comment (cyrildewit) [inline app/ACL/Roles/Providers/DefaultRolesProvider.php:15]: Deze abstractie maakt het mogelijk om op service provider niveau de instance te swappen. Ofwel de S van responsibility (maar 1 reden om te veranderen)

File: app/Console/Commands/ACL/SynchroniseRolesAndPermissions.php [hunk 3]
```diff
@@ -0,0 +1,24 @@
+<?php
+
+declare(strict_types=1);
+
+namespace App\Console\Commands\ACL;
+
+use App\ACL\Contracts\ACLService;
+use Illuminate\Console\Command;
+
+class SynchroniseRolesAndPermissions extends Command
```
Comment (JSTLWK) [inline app/Console/Commands/ACL/SynchroniseRolesAndPermissions.php:10]: Wil je dat ik deze commando ook elke keer bij deployment uitvoer?

**Review by JSTLWK — APPROVED** [review]
Everything looks very clear and easy to understand. I don't have any changes but a note regarding the command.

If we should run it scheduled on the acceptance server.

File: app/Console/Commands/ACL/SynchroniseRolesAndPermissions.php (hunk 3 above)
Comment (cyrildewit) [reply app/Console/Commands/ACL/SynchroniseRolesAndPermissions.php:10]: This should indeed be added to the deployment steps of acceptance. I've created a ticket for you: https://github.com/KA-Huis/space-management/issues/91

**Pull request merged**